packages/protocol/lib/compatibility/version.ts

```diff
@@ -105,3 +124,57 @@ export class ContractVersionDelta {
     return deltas.map(d => d.toString()).join('.')
   }
 }
+
+/**
+ * A mapping {contract name => {@link ContractVersionDelta}}.
+ */
+export interface ContractVersionDeltaIndex {
+  [contract: string]: ContractVersionDelta
+}
+
+/**
+ * A mapping {contract name => {@link ContractVersion}}.
+ */
+export interface ContractVersionIndex {
+  [contract: string]: ContractVersion;
+}
+
+/**
+ * A mapping {contract name => {@link ContractVersion}}.
+ */
+export class ContractVersions {
+  static fromArtifacts = async (artifacts: BuildArtifacts): Promise<ContractVersions>=> {
+    const contracts = {}
+
+    await Promise.all(artifacts.listArtifacts().map(async (artifact) => {
+      contracts[artifact.contractName] = await getContractVersion(makeZContract(artifact))
+    }))
+    return new ContractVersions(contracts)
+  }
+
+  constructor(public readonly contracts: ContractVersionIndex) {}
+}
+
+async function getContractVersion(contract: ZContract): Promise<ContractVersion> {
+  const vm = new VM()
+  const bytecode = contract.schema.deployedBytecode
+  const data = '0x' + abi.methodID('getVersionNumber', []).toString('hex')
+  const nullAddress = '0000000000000000000000000000000000000000'
+  // Artificially link all libraries to the null address.
+  const linkedBytecode = bytecode.split(/[_]+[A-Za-z0-9]+[_]+/).join(nullAddress)
+  const result = await vm.runCall({
+    to: Buffer.from(nullAddress, 'hex'),
+    caller: Buffer.from(nullAddress, 'hex'),
+    code: Buffer.from(linkedBytecode.slice(2), 'hex'),
+    static: true,
+    data: Buffer.from(data.slice(2), 'hex')
+  })
+  if (result.execResult.exceptionError === undefined) {
+    const value = result.execResult.returnValue
+    if (value.length === 4 * 32) {
+      return ContractVersion.fromBuffer(value)
+    }
+  }
+  // If we can't fetch the version number, assume version 1.0.0.0.
```

Comment:
I would manage the error in a different way here:

if the method is not present, if it doesn't exist, then we could assume a default value. (the only thing that worries me about that and the default value, is if someone adds the method with a typo or different signature and doesn't realize it)

if the method exists, but there is an error in calling it, then I would just throw a big exception. It's ok to break there. What do you think?

Reply:
The `ICeloVersionedContract` interface should protect us from this

Reply:
True. Still it may make sense to differentiate between a 'error calling the function' and 'function doesn't exist'.